File: context_menu/combine_linked.py
import os
import sys
import re
import json
from pathlib import Path

# Setup Project Paths
current_dir = Path(__file__).parent.absolute()
project_root = current_dir.parent
# Add both project root AND context_menu to path to ensure we can find pyRitoFile wherever it is
if str(project_root) not in sys.path:
    sys.path.insert(0, str(project_root))
if str(current_dir) not in sys.path:
    sys.path.insert(0, str(current_dir))

try:
    import pyRitoFile
    from pyRitoFile.bin import BIN, BINHasher
    try:
        from pyRitoFile.wad import WADHasher
    except ImportError:
        print("Warning: Could not import WADHasher. Hashed bin matching will be disabled.")
        WADHasher = None
except ImportError:
    print("Error: Could not import pyRitoFile.")
    sys.exit(1)

def get_bin_entry_hashes(bin_obj):
    """Return a set of all entry hashes in a BIN."""
    return {getattr(e, 'hash', '') for e in bin_obj.entries}

def combine_linked():
    if len(sys.argv) < 2: return
    main_bin_path = Path(sys.argv[1]).absolute()
    
    # 1. Determine Skin Root (Parent of 'data')
    root_dir = main_bin_path.parent
    temp = main_bin_path.parent
    while temp.parent != temp:
        if temp.name.lower() == 'data':
            root_dir = temp.parent
            break
        temp = temp.parent

    print(f"--- CONTENT-BASED MERGE ---")
    print(f"Main BIN: {main_bin_path.name}")
    print(f"Scanning Root: {root_dir}")

    # 2. Read Main BIN
    main_bin = BIN().read(str(main_bin_path))
    main_entry_hashes = get_bin_entry_hashes(main_bin)
    
    # 3. Detect champ name to skip base bin
    champ_name = None
    match = re.search(r'/characters/([^/]+)/', main_bin_path.as_posix(), re.IGNORECASE)
    if match: champ_name = match.group(1).lower()

    merged_files = []

    # Load hashed_files.json from root_dir (maps hashed filename -> original path)
    # This is created during WAD extraction and tells us which hashed .bin files map to which paths
    hashed_files_map = {}  # hashed_filename (without .bin) -> original_path
    path_to_hash = {}      # original_path -> hashed_filename (without .bin)

    hashed_json_path = root_dir / "hashed_files.json"
    if hashed_json_path.exists():
        try:
            with open(hashed_json_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
                for hashed_name, orig_path in data.items():
                    # hashed_name is like "38bfb361f6e60b7b.bin"
                    # orig_path is like "data/akali_skins_skin0_skins_skin1_...bin"
                    h = hashed_name.replace('.bin', '').lower()
                    p = orig_path.lower().replace('\\', '/')
                    hashed_files_map[h] = p
                    path_to_hash[p] = h
                print(f"  Loaded {len(data)} entries from hashed_files.json")
        except Exception as e:
            print(f"  Failed to load hashed_files.json: {e}")
    else:
        print("  No hashed_files.json found (files may not be hashed)")

    # 4. Find bins via LINKS (Smart Scan)
    # ... (Keep existing Link check logic) ...
    files_to_process = []
    processed_paths = set()
    
    # Helper to check if a file is the base bin
    def is_base_bin(p):
        return champ_name and p.name.lower() == f"{champ_name}.bin"

    # Helper to check if file is main bin
    def is_main_bin(p):
        return p.resolve() == main_bin_path.resolve()

    # 4a. Check explicitly linked files
    print(f"\n--- LINKED BINS ({len(main_bin.links)} links) ---")
    for link in main_bin.links:
        # Skip non-bin links
        if not link.lower().endswith('.bin'):
            continue

        link_name = Path(link).name
        normalized_link = link.lower().replace('\\', '/')

        candidates = [
            (root_dir / link_name, "name"),
            (root_dir / link, "path"),
        ]

        # Try to find hash from loaded hashtables (reverse lookup)
        hashed_name = None
        if normalized_link in path_to_hash:
            hashed_name = path_to_hash[normalized_link] + ".bin"
            candidates.append((root_dir / hashed_name, "hash-lookup"))

        # Fallback: compute hash if WADHasher available
        if WADHasher and not hashed_name:
            hashed_name = WADHasher.raw_to_hex(normalized_link) + ".bin"
            candidates.append((root_dir / hashed_name, "hash-computed"))

        print(f"  Link: {link}")
        if hashed_name:
            print(f"    -> Hash: {hashed_name}")

        found = False
        for cand, method in candidates:
            if cand.exists() and cand.is_file():
                if cand.resolve() not in processed_paths:
                    if not is_base_bin(cand) and not is_main_bin(cand):
                        print(f"    -> Found via {method}: {cand.name}")
                        files_to_process.append((cand, link))
                        processed_paths.add(cand.resolve())
                        found = True
                    break

        if not found and WADHasher:
            print(f"    -> NOT FOUND") 

    links_to_remove = set()

    for f, original_link in files_to_process:
        try:
            # Open the bin
            mystery_bin = BIN().read(str(f))
            
            # Check if this bin has "new" entries we don't have yet
            new_entries = []
            for e in mystery_bin.entries:
                h = getattr(e, 'hash', '')
                if h and h not in main_entry_hashes:
                    new_entries.append(e)
            
            if new_entries:
                print(f"  [MERGE] Found {len(new_entries)} relevant entries in {f.name}")
                if original_link:
                    print(f"          (Matched via link: {original_link})")
                
                main_bin.entries.extend(new_entries)
                # Update our tracking set
                for e in new_entries:
                    main_entry_hashes.add(getattr(e, 'hash', ''))
                
                merged_files.append(f)
                if original_link:
                    links_to_remove.add(original_link)
        except Exception as e:
            print(f"  [Error] Failed to merge {f.name}: {e}")
            continue

    if not merged_files:
        print("No matching content found to merge.")
        return

    # 5. Save and Cleanup
    # Filter links: Keep those NOT merged
    new_links = []
    for link in main_bin.links:
        if link in links_to_remove:
            continue
        is_champ_bin = champ_name and f"{champ_name}.bin" in link.lower()
        is_non_bin = '/' in link and not link.endswith('.bin')
        
        if is_champ_bin or is_non_bin:
            new_links.append(link)
    main_bin.links = new_links
    
    main_bin.write(str(main_bin_path))

    for f in merged_files:
        try: os.remove(str(f))
        except: pass

    print(f"\n[OK] SUCCESS: Combined {len(merged_files)} files into {main_bin_path.name}")

if __name__ == "__main__":
    combine_linked()

File: native/quartz_cli/src/commands/combine_linked.rs
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::{Path, PathBuf};

use xxhash_rust::xxh64::xxh64;

use crate::utils::{find_root_dir, read_bin, write_bin};

/// Compute xxhash64 of a path (same as WADHasher.raw_to_hex).
fn wad_hash(s: &str) -> u64 {
    xxh64(s.as_bytes(), 0)
}

/// Extract champion name from path like "/characters/akali/skins/..."
fn detect_champ_name(path: &Path) -> Option<String> {
    let posix = path.to_string_lossy().replace('\\', "/").to_lowercase();
    let marker = "/characters/";
    let start = posix.find(marker)? + marker.len();
    let rest = &posix[start..];
    let end = rest.find('/')?;
    Some(rest[..end].to_string())
}

pub fn run(bin_path: &Path) -> Result<(), String> {
    let main_bin_path = bin_path.to_path_buf();
    let root_dir = find_root_dir(bin_path);

    eprintln!("--- CONTENT-BASED MERGE ---");
    eprintln!("Main BIN: {}", main_bin_path.file_name().unwrap_or_default().to_string_lossy());
    eprintln!("Scanning Root: {}", root_dir.display());

    // 2. Read Main BIN
    let mut main_bin = read_bin(&main_bin_path)?;
    let mut main_entry_hashes: HashSet<u32> = main_bin.objects.keys().copied().collect();

    // 3. Detect champ name to skip base bin
    let champ_name = detect_champ_name(&main_bin_path);

    // Load hashed_files.json
    let mut path_to_hash: HashMap<String, String> = HashMap::new();
    let hashed_json_path = root_dir.join("hashed_files.json");
    if hashed_json_path.exists() {
        if let Ok(content) = fs::read_to_string(&hashed_json_path) {
            if let Ok(data) = serde_json::from_str::<HashMap<String, String>>(&content) {
                eprintln!("  Loaded {} entries from hashed_files.json", data.len());
                for (hashed_name, orig_path) in &data {
                    let h = hashed_name.replace(".bin", "").to_lowercase();
                    let p = orig_path.to_lowercase().replace('\\', "/");
                    path_to_hash.insert(p, h);
                }
            }
        }
    } else {
        eprintln!("  No hashed_files.json found (files may not be hashed)");
    }

    // Helpers
    let is_base_bin = |p: &Path| -> bool {
        if let Some(ref champ) = champ_name {
            p.file_name()
                .map(|n| n.to_string_lossy().to_lowercase() == format!("{}.bin", champ))
                .unwrap_or(false)
        } else {
            false
        }
    };
    let is_main_bin = |p: &Path| -> bool {
        p.canonicalize().ok() == main_bin_path.canonicalize().ok()
    };

    // 4. Find bins via LINKS
    let links: Vec<String> = main_bin.dependencies.clone();
    eprintln!("\n--- LINKED BINS ({} links) ---", links.len());

    let mut files_to_process: Vec<(PathBuf, String)> = Vec::new();
    let mut processed_paths: HashSet<PathBuf> = HashSet::new();

    for link in &links {
        if !link.to_lowercase().ends_with(".bin") {
            continue;
        }

        let link_name = Path::new(link).file_name().unwrap_or_default().to_string_lossy().to_string();
        let normalized_link = link.to_lowercase().replace('\\', "/");

        let mut candidates: Vec<(PathBuf, &str)> = vec![
            (root_dir.join(&link_name), "name"),
            (root_dir.join(link), "path"),
        ];

        // Try hash lookup from hashed_files.json
        let mut hashed_name: Option<String> = None;
        if let Some(h) = path_to_hash.get(&normalized_link) {
            hashed_name = Some(format!("{}.bin", h));
            candidates.push((root_dir.join(format!("{}.bin", h)), "hash-lookup"));
        }

        // Fallback: compute xxhash64
        if hashed_name.is_none() {
            let computed = format!("{:016x}.bin", wad_hash(&normalized_link));
            candidates.push((root_dir.join(&computed), "hash-computed"));
            hashed_name = Some(computed);
        }

        eprintln!("  Link: {}", link);
        if let Some(ref h) = hashed_name {
            eprintln!("    -> Hash: {}", h);
        }

        let mut found = false;
        for (cand, method) in &candidates {
            if cand.exists() && cand.is_file() {
                let resolved = cand.canonicalize().unwrap_or_else(|_| cand.clone());
                if !processed_paths.contains(&resolved) && !is_base_bin(cand) && !is_main_bin(cand)
                {
                    eprintln!("    -> Found via {}: {}", method, cand.file_name().unwrap_or_default().to_string_lossy());
                    files_to_process.push((cand.clone(), link.clone()));
                    processed_paths.insert(resolved);
                    found = true;
                }
                break;
            }
        }

        if !found {
            eprintln!("    -> NOT FOUND");
        }
    }

    let mut merged_files: Vec<PathBuf> = Vec::new();
    let mut links_to_remove: HashSet<String> = HashSet::new();

    for (f, original_link) in &files_to_process {
        match read_bin(f) {
            Ok(mystery_bin) => {
                let mut new_entries = Vec::new();
                for (&path_hash, obj) in &mystery_bin.objects {
                    if !main_entry_hashes.contains(&path_hash) {
                        new_entries.push((path_hash, obj.clone()));
                    }
                }

                if !new_entries.is_empty() {
                    eprintln!(
                        "  [MERGE] Found {} relevant entries in {}",
                        new_entries.len(),
                        f.file_name().unwrap_or_default().to_string_lossy()
                    );
                    eprintln!("          (Matched via link: {})", original_link);

                    for (path_hash, obj) in new_entries {
                        main_entry_hashes.insert(path_hash);
                        main_bin.objects.insert(path_hash, obj);
                    }
                    merged_files.push(f.clone());
                    links_to_remove.insert(original_link.clone());
                }
            }
            Err(e) => {
                eprintln!("  [Error] Failed to merge {}: {}", f.file_name().unwrap_or_default().to_string_lossy(), e);
            }
        }
    }

    if merged_files.is_empty() {
        eprintln!("No matching content found to merge.");
        return Ok(());
    }

    // 5. Save and Cleanup — filter links
    let mut new_links: Vec<String> = Vec::new();
    for link in &main_bin.dependencies {
        if links_to_remove.contains(link) {
            continue;
        }
        let is_champ_bin = champ_name
            .as_ref()
            .map(|c| link.to_lowercase().contains(&format!("{}.bin", c)))
            .unwrap_or(false);
        let is_non_bin = link.contains('/') && !link.ends_with(".bin");

        if is_champ_bin || is_non_bin {
            new_links.push(link.clone());
        }
    }
    main_bin.dependencies = new_links;

    write_bin(&main_bin_path, &main_bin)?;

    for f in &merged_files {
        if let Err(e) = fs::remove_file(f) {
            eprintln!("  Warning: could not delete {}: {}", f.display(), e);
        }
    }

    eprintln!(
        "\n[OK] SUCCESS: Combined {} files into {}",
        merged_files.len(),
        main_bin_path.file_name().unwrap_or_default().to_string_lossy()
    );

    Ok(())
}
